activation/verify_state.go

```diff
@@ -0,0 +1,88 @@
+package activation
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.uber.org/zap"
+
+	awire "github.com/spacemeshos/go-spacemesh/activation/wire"
+	"github.com/spacemeshos/go-spacemesh/codec"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/malfeasance/wire"
+	"github.com/spacemeshos/go-spacemesh/sql"
+	"github.com/spacemeshos/go-spacemesh/sql/atxs"
+	"github.com/spacemeshos/go-spacemesh/sql/identities"
+)
+
+func CheckPrevATXs(ctx context.Context, logger *zap.Logger, db sql.Executor) error {
+	collisions, err := atxs.PrevATXCollisions(db)
+	if err != nil {
+		return fmt.Errorf("get prev ATX collisions: %w", err)
+	}
+
+	logger.Info("found ATX collisions", zap.Int("count", len(collisions)))
+	count := 0
+	for _, collision := range collisions {
+		select {
+		case <-ctx.Done():
+			// stop on context cancellation
+			return ctx.Err()
+		default:
+		}
+
+		if collision.NodeID1 != collision.NodeID2 {
+			logger.Panic(
+				"unexpected collision",
+				log.ZShortStringer("NodeID1", collision.NodeID1),
+				log.ZShortStringer("NodeID2", collision.NodeID2),
+				log.ZShortStringer("ATX1", collision.ATX1),
+				log.ZShortStringer("ATX2", collision.ATX2),
+			)
+		}
+
+		malicious, err := identities.IsMalicious(db, collision.NodeID1)
+		if err != nil {
+			return fmt.Errorf("get malicious status: %w", err)
+		}
+
+		if malicious {
+			// already malicious no need to generate proof
+			continue
+		}
+
+		var blob sql.Blob
+		var atx1 awire.ActivationTxV1
+		if _, err := atxs.LoadBlob(ctx, db, collision.ATX1.Bytes(), &blob); err != nil {
+			return fmt.Errorf("get blob %s: %w", collision.ATX1.ShortString(), err)
+		}
+		codec.MustDecode(blob.Bytes, &atx1)
```

Comment:
Please, leave a note that the code should handle the returned `version` (the first value returned from `atxs.LoadBlob()`) or check the version and fail if not V1:
```suggestion
		var atx1 awire.ActivationTxV1
		version,  err := atxs.LoadBlob(ctx, db, collision.ATX1.Bytes(), &blob)
		if err != nil {
			return fmt.Errorf("get blob %s: %w", collision.ATX1.ShortString(), err)
		}
		if version != types.AtxV1 {
			// TODO: support ATX V2
			return errors.New("ATX V1 is not supported yet")
		}
		codec.MustDecode(blob.Bytes, &atx1)
```

Reply:
This code will be removed before we have an ATX v2

Reply:
Added comment explaining why no check for version is needed